chia/wallet/wallet.py

```diff
@@ -332,7 +333,10 @@ async def _generate_unsigned_transaction(
                 raise ValueError(f"Can't send more than {max_send} in a single transaction")
 
         if coins is None:
-            coins = await self.select_coins(total_amount)
+            exclude_coins_list: Optional[List[Coin]] = None
+            if exclude_coins is not None:
```

Comment:
The first line of the function should be something like:
```
if exclude_coins is: None: exclude_coins = []
if coins is None: coins = []
```
That way, there is no ambiguity about things being None or empty. 

I think it's reasonable to be able to use both, but this should be tested explicitly, and if there is a conflict, it should throw an error. Ignoring what the user passed in is a bad idea and will lead to bugs and complaints.

Reply:
My reason for only using `exclude_coins` when selecting coins is that when a user wants to select specific coins via `coins`, he can already filter those client side, as the inputs for both are the same: `Coin`s. That being said, we can ofc also do a server side filter of `coins` by `exclude_coins` if both are set.

When also removing the optional part of both options we will have to throw early in case the server side filter leaves no coins in the array, because otherwise we'd fallback to auto selecting, because now we only check for empty.

> and if there is a conflict, it should throw an error

what do you mean by that, i mean how would they conflict?

Reply:
I believe the suggestion is talking about what happens when a non-empty set of `coins` and a non-empty set of `exclude_coins` are both passed in. This PR ignores in this case the `exclude_coins` which I think is what Mariano is talking about. But if you don't ignore it, then what happens if the same coin is in both `coins` and `exclude_coins` - to which the suggestion was to raise an exception. Somewhat edge case, also very easy to check in python since they are both Sets `bool(coins & excluded_coins)`

Reply:
I say after making it a list, we do if `coins is not None and excluded_coins is not None:` make a copy of coins & a new list and loop through the list only including the non excluded coins. ex:
`old_coins = coins`
`coins: List[Coin] = [] for coin in old_coins:`
       `    if coin is not in excluded_coins:`
        `coins.append(coin)`

Reply:
@emlowe Is the expected behavior of the method that when passing both in we do a filter of the `coins` using `exclude_coins` first thing? in that case we wouldnt need to throw as well. If no, then what is the expected behavior in this scenario?

Reply:
The following piece of code could be added below the coin selection condition:
```
elif exclude_coins is not None:
    if bool(coins & exclude_coins):
        raise ValueError(f"Can't exclude coins specifically included")
    coins = coins - exclude_coins
```
Depending on what the expected behavior of the method is when given both coins and exclude_coins we can also remove the throw before the filter

Reply:
I would remove the throw but put a warning saying that this is not recommended.

Reply:
> I would remove the throw but put a warning saying that this is not recommended.

But if its the expected behavior why warn?

Reply:
> > I would remove the throw but put a warning saying that this is not recommended.
> 
> But if its the expected behavior why warn?

fair point, i think it should error out then.

Reply:
Yes, i agree. So the expected behavior is that when passing both and they intersect we throw. One might even throw directly, even if they dont intersect, because there is no valid case where they are ever used. What do you think?

Reply:
Updated the PR to reflect that :+1: